sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -85,6 +85,25 @@ case class AdaptiveSparkPlanExec(
       case _ => logDebug(_)
     }
 
+  @transient private var ruleContext = new AdaptiveRuleContext(
+    isSubquery = isSubquery,
+    isFinalStage = false)
+
+  private def withRuleContext[T](f: => T): T =
+    AdaptiveRuleContext.withRuleContext(ruleContext) { f }
+
+  private def applyPhysicalRulesWithRuleContext(
```

Comment:
Shall we update `applyPhysicalRules` directly? Do we expect people to call `applyPhysicalRules` without rule context?

Reply:
`applyPhysicalRules` is a kind of private method, not sure how can people use it.

Reply:
I mean, people who develop Spark. When `applyPhysicalRules` should be called instead of `applyPhysicalRulesWithRuleContext`? If never, why do we still keep it?

Reply:
`applyPhysicalRules` it is used by `InsertAdaptiveSparkPlan` for Spark internal rules and at that time we do not have a `AdaptiveSparkPlanExe`. It does not affect user-specifed rules, so I leave it.

Reply:
ah I see